fabtests/examples/example_rdm.c

```diff
@@ -0,0 +1,507 @@
+	 * 	      struct fid_cq **cq, void *context);
+	 */
+	cq_attr.size = 128;
+	cq_attr.format = FI_CQ_FORMAT_MSG;
+	ret = fi_cq_open(domain, &cq_attr, &cq, NULL);
+	if (ret) {
+		printf("fi_cq_open error (%d)\n", ret);
+		return ret;
+	}
+
+	/*
+	 * Bind our CQ to our endpoint to track any sends and receives that
+	 * come in or out on that endpoint. A CQ can be bound to multiple
+	 * endpoints but one EP can only have one send CQ and one receive CQ
+	 * (which can be the same CQ).
+	 */
+
+	/* fi_ep_bind(struct fid_ep *ep, struct fid *fid, uint64_t flags); */
+	ret = fi_ep_bind(ep, &cq->fid, FI_SEND | FI_RECV);
+	if (ret) {
+		printf("fi_ep_bind cq error (%d)\n", ret);
+		return ret;
+	}
+
+	/*
+	 * Initialize our address vector. Address vectors are used to map
+	 * higher level addresses, which may be more natural for an application
+	 * to use, into fabric specific addresses. An AV_TABLE av will map
+	 * thse addresses to indexed addresses, starting with fi_addr 0. These
+	 * addresses are used in data transfer calls to specify which peer to
+	 * send to/recv from. Address vectors are only used for FI_EP_RDM and
+	 * FI_EP_DGRAM endpoints, allowing the application to avoid connection
+	 * management. For FI_EP_MSG endpoints, the AV is replaced by the
+	 * traditional TCP listen/connect/accept steps.
+	 */
+
+	/*
+	 * fi_av_open(struct fid_domain *domain, struct fi_av_attr *attr,
+	 * 	      struct fid_av **av, void *context);
+	 */
+	av_attr.type = FI_AV_TABLE;
+	av_attr.count = 1;
+	ret = fi_av_open(domain, &av_attr, &av, NULL);
+	if (ret) {
+		printf("fi_av_open error (%d)\n", ret);
+		return ret;
+	}
+
+	if (dst_addr) {
+		/*
+		 * Here the client inserts the server's address into the AV
+		 * which returns an fi_addr to use when sending data to the
+		 * peer (server). Note that only the client has to insert the
+		 * server's address into its AV since it is the one sending.
+		 * The server does not need to have a peer's address in its AV
+		 * in order to receive a message.
+		 */
+
+		/*
+		 * fi_av_insert(struct fid_av *av, void *addr, size_t count,
+		 * 		fi_addr_t *fi_addr, uint64_t flags,
+		 * 		void *context);
+		 */
+		ret = fi_av_insert(av, info->dest_addr, 1, &fi_addr, 0, NULL);
+		if (ret != 1) {
+			printf("fi_av_insert error (%d)\n", ret);
+			return ret ? ret : -1;
+		}
+	}
+
+	/*
+	 * Bind the AV to the EP. The EP can only send data to a peer in its
+	 * AV.
+	 */
+
+	/* fi_ep_bind(struct fid_ep *ep, struct fid *fid, uint64_t flags); */
+	ret = fi_ep_bind(ep, &av->fid, 0);
+	if (ret) {
+		printf("fi_ep_bind av error (%d)\n", ret);
+		return ret;
+	}
+
+	/*
+	 * Once we have all our resources initialized and ready to go, we can
+	 * enable our EP in order to send/receive data.
+	 */
+
+	/* fi_enable(struct fid_ep *ep); */
+	ret = fi_enable(ep);
+	if (ret) {
+		printf("fi_enable error (%d)\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+/* All OFI resources are cleaned up using the same fi_close(fid) call. */
```

Comment:
Expand comment to indicate that resources must be closed in a specific order to allow references between objects to be removed correctly.  E.g. the EP must be closed before the CQ or AV.